packages/plugin-dcap/src/actions/on-chain.ts

```diff
@@ -0,0 +1,111 @@
+import {
+    getDCAPMode,
+    getTEEMode,
+    hasPrivateKey,
+    hasTEEMode,
+} from "../utils.js";
+
+export const dcapOnChainVerifyAction: Action = {
+    name: "DCAP_ON_CHAIN",
+    description:
+        "This plugin is used to generate DCAP attestation and verify it on-chain. The user can also use the keyword DCAP_ON_CHAIN to trigger this action.",
+    similes: [
+        "DCAP",
+        "DCAP_ATTESTATION",
+        "DCAP_TEE",
+        "DCAP_SGX",
+        "DCAP_TDX",
+        "VERIFY_ATTESTATION",
+        "VERIFY_DCAP",
+        "DCAP_VERIFICATION",
+        "ATTESTATION",
+        "GENERATE_ATTESTATION",
+    ],
+    examples: [
+        [
+            {
+                user: "{{user1}}",
+                content: {
+                    text: "Generate a DCAP attestation and verify it on-chain",
+                },
+            },
+            {
+                user: "{{user2}}",
+                content: {
+                    text: "Of course, hanlding it now...",
+                    action: "DCAP_ON_CHAIN",
+                },
+            },
+        ],
+        [
+            {
+                user: "{{user1}}",
+                content: { text: "Verify the DCAP attestation on-chain" },
+            },
+            {
+                user: "{{user2}}",
+                content: {
+                    text: "Of course, hanlding it now...",
+                    action: "DCAP_ON_CHAIN",
+                },
+            },
+        ],
+        [
+            {
+                user: "{{user1}}",
+                content: { text: "DCAP_ON_CHAIN" },
+            },
+            {
+                user: "{{user2}}",
+                content: {
+                    text: "Of course, hanlding it now...",
+                    action: "DCAP_ON_CHAIN",
+                },
+            },
+        ],
+    ],
+    async validate(runtime, message) {
+        if (!hasPrivateKey(runtime)) return false;
+        const mode = getDCAPMode(runtime);
+        if (!mode) return false;
+        if (mode === DCAPMode.PLUGIN_TEE) return hasTEEMode(runtime);
+        return true;
+    },
+    async handler(runtime, message, state, options, callback) {
+        const { agentId } = runtime;
+        const { userId, roomId, content } = message;
+        const quote = await getQuote(
+            // Attestation will be generated based on the message info
+            JSON.stringify({
+                agentId,
+                timestamp: Date.now(),
+                message: { userId, roomId, content: content.text },
+            }),
+            getDCAPMode(runtime),
+            getTEEMode(runtime)
+        );
+
+        const reply = (text: string) =>
+            callback({
+                text,
+                // source: quote,
+                action: "DCAP_ON_CHAIN",
+            });
+        try {
+            const tx = await verifyAndAttestOnChain(
+                runtime.getSetting("EVM_PRIVATE_KEY")!,
```

Comment:
Can this be interchangeable? If this is running in TEE Cloud, the agent could derive its key in the TEE & be able to remove the need to store the private key in the env variables.

Reply:
Hey @HashWarlock, we originally just wanted to obtain the key from the setting with the <env_var> used as an abstraction for the key so we don't want to make any assumptions where the key comes from, e.g., HSMs / SVSMs / KMSs / etc or any plugins.

That said, if the general consensus is that plugin-evm must be included then we could change it to obtain the key using the WalletProvider instead.

Reply:
I think it's fine. It would be more of scope utility function that can grab the key from plugin tee or the env vars. We can always make the change if its requested by a dev in the future.